Document the tcp_queue_length core check (#6977)
DataDog/integrations-core: tcp_queue_length/README.md

# Agent Check: TCP Queue Length

## Overview

> This Check monitors the size of the Linux TCP receive and send queues.

Reviewer: ```suggestion
This check monitors the size of the Linux TCP receive and send queues.
```

## Setup

### Installation

`tcp_queue_length` is a core Agent 6/7 check that rely on an eBPF part implemented in `system-probe`.

> The eBPF program used by `system-probe` is compiled at runtime and requires to have access to the proper kernel headers.

Reviewer: ```suggestion
The eBPF program used by `system-probe` is compiled at runtime and requires access to the proper kernel headers.
```

On Debian-like distribution, you should install the kernel headers like this:
```sh
apt install -y linux-headers-$(uname -r)
```

On RHEL-like distribution, you should install the kernel headers like this:
```sh
yum install -y kernel-headers-$(uname -r)
```

### Configuration

> Enabling it requires to enable it both for `system-probe` and the core agent.

Reviewer: ```suggestion
Enabling the `tcp_queue_length` integration requires both the `system-probe` and the core agent to have the configuration option enabled.
```

Inside the `system-probe.yaml` configuration file, the following parameter must be set:
```yaml

> system_probe_config:

Reviewer 2: should we add `enabled: true` ?

Author: Definitively !

  enable_tcp_queue_length: true
```

> For the “core” agent, a `tcp_queue_length.d/conf.yaml` file is needed in the `conf.d` directory [with the following content][1].

Reviewer: ```suggestion
1. Edit the `tcp_queue_length.d/conf.yaml` file, in the `conf.d/` folder at the root of your
   Agent's configuration directory to start collecting your tcp_queue_length performance data.
   See the [sample tcp_queue_length.d/conf.yaml][1] for all available configuration options.

2. [Restart the Agent][3].
```

The `only_count_nb_contexts` parameter controls whether:
* the check should collect all the data—that is, the size of the TCP queues for every single connection (`false`)
* the check should collect only the number of timeseries it would normally collect (`true`). This is the default.

### Configuration with Helm

> With the [Datadog Helm chart][2], the `datadog.systemProbe.enableTCPQueueLength` parameter is the only thing to set in `values.yaml` to enable the check.

Reviewer 2: should we mention that both the system probe and process agent should be enabled too?

Author: Indeed!
I added the parameter to activate `system-probe`.
The `process-agent`, on the other hand isn’t needed.

### Validation

[Run the Agent's `status` subcommand][3] and look for `tcp_queue_length` under the Checks section.

## Data Collected

### Metrics

`tcp_queue_length` collects, for each TCP connection the minimum and the maximum amount of data in the send and receive buffers between each data collection time point.

See [metadata.csv][4] for a list of metrics provided by this Integration

> 

Reviewer: Missing sections for Events and Service Checks

```suggestion

### Service Checks

The TCP Queue Length check does not include any service checks.

### Events

The TCP Queue Length check does not include any events.

## Troubleshooting

Need help? Contact [Datadog support][5].
```

```
[5]: https://docs.datadoghq.com/help/
```

# Agent Check: TCP Queue Length

## Overview

This check monitors the size of the Linux TCP receive and send queues.

## Setup

### Installation

`tcp_queue_length` is a core Agent 6/7 check that relies on an eBPF part implemented in `system-probe`.

The eBPF program used by `system-probe` is compiled at runtime and requires you to have access to the proper kernel headers.

On Debian-like distributions, install the kernel headers like this:
```sh
apt install -y linux-headers-$(uname -r)
```

On RHEL-like distributions, install the kernel headers like this:
```sh
yum install -y kernel-headers-$(uname -r)
```

### Configuration

Enabling the `tcp_queue_length` integration requires both the `system-probe` and the core agent to have the configuration option enabled.

Inside the `system-probe.yaml` configuration file, the following parameters must be set:
```yaml
system_probe_config:
  enabled: true
  enable_tcp_queue_length: true
```

> 1. Edit the `tcp_queue_length.d/conf.yaml` file, in the `conf.d/` folder at the root of your
>    Agent's configuration directory to start collecting your tcp_queue_length performance data.
>    See the [sample tcp_queue_length.d/conf.yaml][1] for all available configuration options.
> 
> 2. [Restart the Agent][3].
> 
> The `only_count_nb_contexts` parameter controls whether:
> * the check should collect all the data—that is, the size of the TCP queues for every single connection (`false`)
> * the check should collect only the number of timeseries it would normally collect (`true`). This is the default.

Reviewer: ```suggestion
1. Edit the `tcp_queue_length.d/conf.yaml` file, in the `conf.d/` folder at the root of your
   Agent's configuration directory to start collecting your tcp_queue_length performance data.
   See the [sample tcp_queue_length.d/conf.yaml][1] for all available configuration options.

    The `only_count_nb_contexts` parameter is enabled by default and collects only the number of timeseries it would normally collect. If disabled, the check will collect all data-- that is, the size of the TCP queues for every single connection.

2. [Restart the Agent][3].

```

I think formatting the option like this might be easier to read and understand it's referencing the configuration file.

Reviewer: Sorry the diff looks like a lot changed, but I just reworded the section about `only_count_nd_contexts` and moved it after Step 1.
